.github/workflows/cmake.yml

```diff
@@ -86,33 +69,68 @@ jobs:
+      if: ${{ matrix.debug-macros == 'ALL_DEBUG' }}
     - name: Create build environment
       working-directory: ${{ github.workspace }}
       # Note that this needs to run *even if* the Toolchain was built,
       # in order to set options like BUILD_LAGOM.
       run: |
         mkdir -p Build
         cd Build
-        cmake .. -GNinja -DBUILD_LAGOM=ON -DENABLE_ALL_THE_DEBUG_MACROS=ON -DCMAKE_C_COMPILER=gcc-10 -DCMAKE_CXX_COMPILER=g++-10
+        cmake .. -GNinja -DBUILD_LAGOM=ON -DCMAKE_C_COMPILER=gcc-10 -DCMAKE_CXX_COMPILER=g++-10
+      if: ${{ matrix.debug-macros == 'NORMAL_DEBUG' }}
 
-    # === ACTUALLY BUILD AND TEST ===
+    # === ACTUALLY BUILD ===
 
     - name: Build Serenity and Tests
       working-directory: ${{ github.workspace }}/Build
       run: cmake --build .
     - name: Lint (Phase 2/2)
       working-directory: ${{ github.workspace }}/Meta
       run: ./check-symbols.sh
-    - name: Run CMake tests
+
+    - name: Create Serenity Rootfs
+      if: ${{ matrix.debug-macros == 'NORMAL_DEBUG'}}
       working-directory: ${{ github.workspace }}/Build
-      run: CTEST_OUTPUT_ON_FAILURE=1 ninja test || ${{ matrix.allow-test-failure }}
-      timeout-minutes: 2
-    - name: Run JS tests
-      working-directory: ${{ github.workspace }}/Build/Meta/Lagom
-      run: DISABLE_DBG_OUTPUT=1 ./test-js || ${{ matrix.allow-test-failure }}
-    - name: Run LibCompress tests
-      working-directory: ${{ github.workspace }}/Build/Meta/Lagom
-      run: ./test-compress
+      run: |
+        echo -e "\n[TestRunner@ttyS0]\n" >> ../Base/etc/SystemServer.ini
+        echo -e "Executable=/home/anon/tests/run-tests-and-shutdown.sh\n" >> ../Base/etc/SystemServer.ini
+        echo -e "StdIO=/dev/ttyS0\n" >> ../Base/etc/SystemServer.ini
+        echo -e "Environment=DO_SHUTDOWN_AFTER_TESTS=1 TERM=xterm PATH=/bin:/usr/bin:/usr/local/bin\n" >> ../Base/etc/SystemServer.ini
+        echo -e "User=anon\n" >> ../Base/etc/SystemServer.ini
+        echo -e "WorkingDirectory=/home/anon\n" >> ../Base/etc/SystemServer.ini
+        echo -e "BootModes=text\n" >> ../Base/etc/SystemServer.ini
+        ninja install && ninja image
+
+    - name: Run On-Target Tests
+      if: ${{ matrix.debug-macros == 'NORMAL_DEBUG'}}
+      working-directory: ${{ github.workspace }}/Build
+      env:
+        SERENITY_KERNEL_CMDLINE: "boot_mode=text"
+        SERENITY_RUN: "ci"
+      run: ninja run
+      timeout-minutes: 10
+    - name: Print target logs
+      if: ${{ !cancelled() && matrix.debug-macros == 'NORMAL_DEBUG'}}
+      working-directory: ${{ github.workspace }}/Build
+      run: cat ./debug.log
+
+
+  notify_irc:
+    needs: [build_and_test_serenity, build_and_test_lagom]
+    runs-on: ubuntu-20.04
+    if: always() && github.repository == 'SerenityOS/serenity' && (github.event_name == 'pull_request' || (github.event_name == 'push' && github.ref == 'refs/heads/master'))
```

Comment:
`always() && ` is redundant.

Have you replaced `github.repository == 'SerenityOS/serenity'` with your own repository name (and your own separate channel in `notify_irc.py`) to test whether this works? I never used the `needs:` feature before, and you reported that "not working as [you] hoped". Well, what does it do? What did your elf eyes see? :P

Reply:
Correction: [`always() &&` is apparently not redundant](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-not-requiring-dependent-jobs-to-be-successful)

Reply:
I have not properly tested the IRC notifications with a real IRC server, you're right. It seems that for the matrix jobs, whichever one completes last is the one that gets the build status updated here when we check this job.

And yes, always is not redundant :/ . If there's a better way to say "please run this job always and give me a bool that says 'ALL previous jobs [succeeded|failed]'" using github actions I couldn't find one...